### ISSUE TITLE
The qwen image edit2.1 support？

### ISSUE THREAD

**Issue by buxinzi2233** [sorryhyun/anima_lora#102]
https://huggingface.co/Qwen/Qwen-Image-2.1
It’s been a while—apologies for the intrusion at this hour. However, this new Qwen model is truly impressive; it runs locally with just 8GB VRAM and 32GB RAM. I just used it to create a character sheet (three-view layout) based on my avatar, and the results were excellent—generated in a single pass with great consistency and faithful style preservation. A friend of mine who is training it in the cloud mentioned that it converges very quickly, yielding good results for style training after just 2,000 steps. Although its native image generation capabilities are somewhat lacking—bearing a strong "GPT-like" quality—using LoRA combined with an editing workflow allows for style changes while maintaining the core subject elements. Given that it is only a 7B model, would you consider experimenting with it—specifically regarding adaptation training and your SPD sampler?

**Comment by sorryhyun** [comment]
hello, glad to hear. I think 2 things are important factors: 1. does it have japanese anime knowledge like mage, 2. did it observed nsfw images while training like mage. if not, it will require substantive steps to train booru images.

**Comment by sorryhyun** [comment]
okay I just checked it has no problem on the 2 factors I mentioned. I'll test shortly.

**Comment by buxinzi2233** [comment]
Looking forward to hearing good news from you.

**Comment by sorryhyun** [comment]
I tested it. Training itself runs fine on 16 GB: LoRA on the 32 transformer blocks driven by this repo's own block swapper (5 of 32 swapped) plus activation checkpointing. Rank 16, 30 images at native aspect and ~1 MP, 8 epochs = 240 steps in 15.7 min, peak 13.96 GB. The text encoder (17.5 GB) and the VAE are precached, so training never loads them. At 1024 the step is compute bound rather than PCIe bound, so per block torch.compile measured no gain.

Two findings that matter for booru datasets:

1. Danbooru tags are not trained at all. Qwen3-VL is a natural language encoder, so it reads a tag string as ordinary words, not as the booru tagging convention. Rating prefixes are completely inert: safe / sensitive / nsfw constrain nothing. A tag bag with no clothing tags ("1girl, solo, upper body, portrait, simple background") renders nude in the base arm with no LoRA attached, because the model is just filling in an underspecified sentence. Style is unstable inside a single prompt set as well, with adjacent prompts coming out anime and photoreal. Whatever works is the plain English sitting inside the tags, not the tag system.

2. It is not converging in small steps. 240 steps gave only a modest style shift. Your friend's 2000 steps sounds much closer to the real cost, at least for booru style data where the caption format is off distribution to begin with.

No SPD work yet.

**Comment by buxinzi2233** [comment]
Indeed, it seems the Qwen team didn't specifically train the model using "booru" formatting. A friend of mine trained a LoRA using natural language prompts; by loading it in "edit mode" and using the prompt "change <image1> style to anime style," he successfully shifted the art style while preserving the composition and elements. It might yield even better results to include tags like "anime style" or "XXX style" during training. That said, I think achieving a LoRA with such a noticeable effect after just 2,000 steps is already excellent.

**Comment by buxinzi2233** [comment]
Additionally, the Qwen team mentioned that they specifically fine-tuned two models and prepared two sets of system prompts to optimize user prompts; the relevant repository is https://github.com/QwenLM/Qwen-Image-2.1/tree/main.

**Comment by buxinzi2233** [comment]
However, these two fine-tuned models themselves come with censorship.

**Comment by sorryhyun** [comment]
I see. Anyway I will adopt the model to be lora-trainable in optimized way, incoming v2.0.0.beta2 release, which will be done umm... in this week. It will be pre-release though. Thanks for recommendation anyway.

**Comment by buxinzi2233** [comment]
No, no, no—the pleasure is mine; thank you for bringing this efficient trainer to the community.

**Comment by sorryhyun** [comment]
would you try https://github.com/sorryhyun/anima_lora/tree/qwen21 ?

**Comment by buxinzi2233** [comment]
I'll give it a try as soon as possible.

**Comment by buxinzi2233** [comment]
https://github.com/LingyeSoul/MonadForge
I trained Anima on a different project branch because FP16 protection was implemented there. Since my own device only has 8GB of VRAM, I’m going to see if I can train it; meanwhile, I’ve asked a friend with a 5090 to run some tests as well.

**Comment by buxinzi2233** [comment]
Alright, I’ve successfully started the training job, and everything looks normal so far; I’m using a laptop equipped with an RTX 4070 GPU. Surprisingly, the execution speed is quite fast. Judging by the current status, the speed doesn't seem to have peaked yet and is still slowly climbing. However, there’s one thing I’m not sure about: is FlashAttention enabled by default?

<img width="1263" height="1366" alt="Image" src="https://github.com/user-attachments/assets/74c9065e-5a14-44af-a156-737d04fb75f9" />

**Comment by sorryhyun** [comment]
yes, flash attention is provided and automatically being used. I think you're on right speed since 5070ti also made 4s/step

**Comment by buxinzi2233** [comment]
Okay, I’ve finished the training run. The first takeaway is that the speed was faster than expected. Secondly, while LoRA works, a learning rate of 1e-4 clearly isn't quite sufficient for this model; after 888 steps, it seems to have had some impact, but not much. A rate of 2e-4 might be more appropriate.

**Comment by buxinzi2233** [comment]
> yes, flash attention is provided and automatically being used. I think you're on right speed since 5070ti also made 4s/step

In actual operation, the final speed is approximately 13 seconds per step; I find it quite impressive to achieve such speed when training a 7B model on a GPU with 8GB of VRAM.

**Comment by buxinzi2233** [comment]
Oh, by the way, I wanted to ask if you might be interested in the fp16 protection implemented in the https://github.com/LingyeSoul/MonadForge project. While the performance might not be the absolute best, it enables training on older GPUs that lack bf16 support, and the speed feels quite decent; on a V100 (16GB) at 1024 resolution—without block swapper or activation checkpointing—it uses around 13GB of VRAM with a speed of 1.1s/step. However, training at 1536 resolution might not be feasible; it tends to result in NaNs after a thousand-plus steps (though this might have been fixed, I haven't looked into it closely).

**Comment by sorryhyun** [comment]
That's nice. Unfortunately since fp16 is quite lossy compared to bf16, we're not supporting the feature, though you guys have a workarounds `--mixed_precision fp16` in main trainer loop.

**Comment by buxinzi2233** [comment]
> That's nice. Unfortunately since fp16 is quite lossy compared to bf16, we're not supporting the feature, though you guys have a workarounds `--mixed_precision fp16` in main trainer loop.

Okay, thanks for the suggestion; I'll take a look. Also, thanks for the Qwen compatibility work.

**Comment by sorryhyun** [comment]
released as v2.0.0.beta2 prerelease
